sound/soc/sof/ipc4-pcm.c

```diff
@@ -495,17 +519,51 @@ static int sof_ipc4_pcm_dai_link_fixup(struct snd_soc_pcm_runtime *rtd,
 				use_chain_dma = true;
 		}
 	}
+
+	fe_rate = params_rate(params);
+
 	/*
-	 * Always set BE format to 32-bits for both playback and capture, except if chained
-	 * DMA is in use. In chained DMA case the back-end and front-end format must match.
+	 * Set audio format and rate mask based on available formats
+	 * defined for the DAI/BE copier. Copiers are not used with
+	 * chain_dma, so must be skipped for this case. Copiers can
+	 * handle arbitrary number of channels, so channel count is
+	 * not limited here.
 	 */
 	if (!use_chain_dma) {
-		snd_mask_none(fmt);
-		snd_mask_set_format(fmt, SNDRV_PCM_FORMAT_S32_LE);
+		struct sof_ipc4_available_audio_format *available_fmt = &ipc4_copier->available_fmt;
+		int i;
+
+		snd_mask_none(fmt_mask);
+
+		for (i = 0; i < available_fmt->audio_fmt_num; i++) {
+			struct sof_ipc4_audio_format *fmt =
+				&available_fmt->base_config[i].audio_fmt;
+			unsigned int bit_depth;
+
+			bit_depth = SOF_IPC4_AUDIO_FORMAT_CFG_V_BIT_DEPTH(fmt->fmt_cfg);
+			if (ipc4_set_fmt_mask(fmt_mask, bit_depth) < 0)
+				return -EINVAL;
```

Comment:
Should we return the return value of `ipc4_set_fmt_mask`? Although it is -EINVAL, too.

Reply:
Ack, fixed (this code removed based on the other discussion on need to modify format mask in dai_fixup).